src/base/ClpMap.h

```diff
@@ -277,8 +277,9 @@ template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
 ClpMap<Key, Value, MemoryUsedBy>::Entry::Entry(const Key &aKey, const Value &v, const Ttl ttl) :
     key(aKey),
     value(v),
-    expires(Sum(squid_curtime, ttl).value_or(std::numeric_limits<time_t>::max()))
+    expires(0) // reset below
 {
+    SetToNaturalSumOrMax(expires, squid_curtime, ttl);
```

Comment:
I do not like this variable setting syntax because it does not look like an assignment and would look weird (and may cause warnings) if used for data member initialization (where the PR currently uses 0 with a "reset below" comment). 

The problem is difficult to fix because we do not want the developer to guess the variable type. In other words, we do **not** want this:

```C++
expires = NaturalSum<long>(squid_curtime, ttl); // XXX: type guessing
```

We could force developers to use decltype, but that feels a bit cruel, might cause const-related warnings when used in initialization lists, and duplicates code:

```C++
expires = NaturalSum<decltype(expires)>(squid_curtime, ttl); // XXX: code duplication; warnings?
```

I was considering adding a dummy/unused char or c-string parameter to stuff a visual assignment clue:

```C++
SetToNaturalSumOrMax(expires, '=', squid_curtime, ttl); // XXX: Too weird?
```

@squid-cache/commiters, @eduard-bagdasaryan, any better ideas?

Reply:
How about just replacing the Sum(x,y) with NaturalSum(x,y) instead of trying to re-write the entire line of code?
That removes the need for whole global template function just for this one line of code.

Reply:
> How about just replacing the Sum(x,y) with NaturalSum(x,y)

That replacement must not compile: One cannot just replace Sum(a,b) with NaturalSum(a,b) because, unlike the buggy Sum(), NaturalSum() does not try to guess the type of the stored result by the type of the first function argument. That first argument may have nothing to do with the stored result type and, hence, the subsequent assignment (not the sum) may overflow. This is one of the bugs NaturalSum() fixes and prevents.

There are solutions that would compile, but they have other problems as discussed in my [comment](https://github.com/squid-cache/squid/pull/869/files#r678710650).

Reply:
Then we cannot use NaturalSum at all for this initialization. The size of time_t may be smaller type than int/Ttl and its max limit varies depending on environment locale - which can make it much smaller than both types.

IIRC, Sum was designed to rely on type conversion rules in order to avoid this type of limitation.

Reply:
> The size of time_t may be smaller type than int/Ttl and its max limit varies depending on environment locale - which can make it much smaller than both types.

Let's assume that is true.

> Then we cannot use NaturalSum at all for this initialization. 

The proposed initialization code effectively uses NaturalSum<time_t>() (without dangerously hard-coding the current `expires` type). That NaturalSum() specialization is supposed to take time_t limits and all the other factors you mentioned into account. If you find any bugs, please point them out!
 
> IIRC, Sum was designed to rely on type conversion rules in order to avoid this type of limitation.

IIRC, your recollection is wrong. Sum() designers were simply unaware of some of the problems they were supposed to solve and, hence, of the bugs their Sum() implementation was introducing. NaturalSum() is now fixing those bugs.

N.B. C++ type conversions do not protect from overflows.